File: src/main/scala/conway/Game.scala
package conway

import scala.collection.immutable
import scala.collection.mutable

/**
 * A Conway's Game of Life coordinator, capable of dealing with boards of size rows x columns.
 */
class Game(rows: Int, columns: Int) {

	private val neighborOffsets: Set[(Int, Int)] = immutable.Set(
		(-1, -1), (-1, 0), (-1, 1),
		(0, -1),           (0, 1),
		(1, -1),  (1, 0),  (1, 1)   
	)

	private val cells: mutable.Set[Point] = mutable.Set.empty

	for (row <- 0 until rows) {
		for (column <- 0 until columns) {
			cells += Point(row, column)
		}
	}

	def turn(board: Board): Board = {
		// Select creatures that should die
		val deaths = board.creatures.filter(c => shouldDie(neighborCount(board, c)))

		// Create newly born creatures where an empty cell has exactly 3 neighbors
		val births = cells
			.filter(cell => !board.containsCreature(cell) && shouldSpawn(neighborCount(board, cell)))
			.foldLeft(mutable.Set.empty[Creature])(
				(creatures, cell) => creatures += Creature.at(cell)
			)

		// Create a new board with same dimensions, minus deaths, plus births
		Board(board.rows, board.columns, board.creatures.++(births).--(deaths))
	}

	private def neighborCount(board: Board, c: Coordinate): Int = {
		neighborOffsets.count(offset => board.containsCreature(Point(c.row + offset._1, c.column + offset._2)))
	}

	private def shouldDie(neighbors: Int): Boolean = neighbors <= 1 || neighbors >= 4

	private def shouldSpawn(neighbors: Int): Boolean = neighbors == 3
}

File: java8/src/main/java/conway/java/Game.java
package conway.java;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.Sets;

import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

public class Game {

	private static final List<Offset> NEIGHBOR_OFFSETS = ImmutableList.of(
			new Offset(-1, -1), new Offset(-1, 0), new Offset(-1, 1),
			new Offset(0, -1),                     new Offset(0, 1),
			new Offset(1, -1),  new Offset(1, 0),  new Offset(1, 1)
	);

	private final Set<Point> cells = new HashSet<>();

	public Game(int rows, int columns) {
		for (int row = 0; row < rows; row++) {
			for (int column = 0; column < columns; column++) {
				cells.add(new Point(row, column));
			}
		}
	}

	public Board turn(Board board) {
		// Select creatures that should die
		Set<Creature> deaths = board.getCreatures().stream()
				.filter(c -> shouldDie(neighborCount(board, c)))
				.collect(Collectors.toSet());

		// Create newly born creatures where an empty cell has exactly 3 neighbors
		Set<Creature> births = cells.parallelStream()
				.filter(cell -> !board.creatureAtCoordinate(cell) && shouldSpawn(neighborCount(board, cell)))
				.map(Creature::at)
				.collect(Collectors.toSet());

		// Create a new board with same dimensions, minus deaths, plus births
		Set<Creature> newCreatures = Sets.union(board.getCreatures(), births).stream()
				.filter(c -> !deaths.contains(c))
				.collect(Collectors.toSet());
		return new Board(board.getRows(), board.getColumns(), newCreatures);
	}

	private long neighborCount(Board board, Coordinate c) {
		return NEIGHBOR_OFFSETS.stream()
				.filter(o -> board.creatureRelativeToCoordinate(c, o))
				.count();
	}

	private boolean shouldDie(long neighbors) {
		return neighbors <= 1 || neighbors >= 4;
	}

	private boolean shouldSpawn(long neighbors) {
		return neighbors == 3;
	}
}
